File: ruby/lib/app_identity/versions.rb
# frozen_string_literal: true

require "time"
require "digest/sha2"
require "securerandom"
require "set"

require_relative "validation"
require_relative "error"

class AppIdentity
  module Versions
    class Base
      class << self
        def defined
          @defined ||= {}
        end

        def instance
          @_instance = new
        end

        def inherited(subclass)
          match = /V(?<version>\d+)\z/.match(subclass.name)
          return unless match
          version = match[:version].to_i

          raise AppIdentity::Error, "version is not unique" if Base.defined.has_key?(version)

          subclass.define_method(:version) { version }

          Base.defined[version] = subclass.instance
        end

        private :new
      end

      def make_digest(raw)
        digest_algorithm.hexdigest(raw).upcase
      end

      def inspect
        v = "version=#{version}" if respond_to?(:version)
        n = "nonce=#{nonce_type}" if respond_to?(:nonce_type)
        d = "digest=#{digest_algorithm}" if respond_to?(:digest_algorithm)

        "#<#{[self.class, v, n, d].compact.join(" ")}>"
      end

      def check_nonce!(nonce, _config)
        raise AppIdentity::Error, "nonce must not be nil" if nonce.nil?
        raise AppIdentity::Error, "nonce must be a string" unless nonce.is_a?(String)
        raise AppIdentity::Error, "nonce must not be blank" if nonce.empty?
        raise AppIdentity::Error, "nonce must not contain colon characters" if /:/.match?(nonce)
      end
    end

    class RandomNonce < Base
      def nonce_type
        :random
      end

      def generate_nonce
        SecureRandom.urlsafe_base64(32)
      end
    end

    class TimestampNonce < Base
      include AppIdentity::Validation

      def nonce_type
        :timestamp
      end

      def generate_nonce
        Time.now.utc.strftime("%Y%m%dT%H%M%S.%6NZ")
      end

      def check_nonce!(nonce, config)
        super(nonce, config)

        timestamp = parse_timestamp!(nonce)

        fuzz = config_fuzz(config)
        diff = (Time.now.utc - timestamp).abs.to_i

        raise AppIdentity::Error, "nonce is invalid" unless diff <= fuzz
      end

      private

      def config_fuzz(config)
        config&.dig(:fuzz) || 600
      end

      def parse_timestamp!(nonce)
        Time.strptime(nonce, "%Y%m%dT%H%M%S.%N%Z").tap { |value| raise if value.nil? }.utc
      rescue
        raise AppIdentity::Error, "nonce does not look like a timestamp"
      end
    end

    # V1 is the original algorithm, using a permanent nonce value and SHA256
    # digests. The use of this version is strongly discouraged for new clients.
    class V1 < RandomNonce
      def digest_algorithm
        Digest::SHA256
      end
    end

    # V2 uses a timestamp-based nonce value with SHA256 digests.
    #
    # The nonce values will be verified to be within plus or minus a configured
    # number of seconds.
    class V2 < TimestampNonce
      def digest_algorithm
        Digest::SHA256
      end
    end

    # V3 uses a timestamp-based nonce value with SHA384 digests.
    #
    # The nonce values will be verified to be within plus or minus a configured
    # number of seconds.
    class V3 < TimestampNonce
      def digest_algorithm
        Digest::SHA384
      end
    end

    # V4 uses a timestamp-based nonce value with SHA512 digests.
    #
    # The nonce values will be verified to be within plus or minus a configured
    # number of seconds.
    class V4 < TimestampNonce
      def digest_algorithm
        Digest::SHA512
      end
    end

    class << self
      include AppIdentity::Validation

      # Looks up the version instance by version.
      def [](version)
        AppIdentity::Versions::Base.defined.fetch(version)
      end

      # Checks to see if the version has been defined.
      def valid?(version)
        AppIdentity::Versions::Base.defined.has_key?(version)
      end

      # Tests that the version is valid or raises an exception.
      def valid!(version)
        return true if valid?(version)
        raise AppIdentity::Error, "version must be one of #{AppIdentity::Versions::Base.defined.keys}"
      end

      # Checks to see if the version is valid and not explicitly disallowed,
      # either in the provided list or in global list.
      def allowed?(version, provided = nil)
        valid?(version) && !disallowed.member?(version) &&
          (provided ? !Set.new(provided).member?(version) : true)
      end

      # Tests that the version is valid and not explicitly disallowed or raises
      # an exception.
      def allowed!(version, provided = nil)
        return true if valid!(version) && allowed?(version, provided)
        raise AppIdentity::Error, "version #{version} has been disallowed"
      end

      # Globally disallow the listed versions.
      def disallow(*versions)
        disallowed.merge(coerce_versions(versions))
      end

      # Globally allow the listed versions.
      def allow(*versions)
        disallowed.subtract(coerce_versions(versions))
      end

      private

      def coerce_versions(versions)
        versions.select { |v| v.is_a?(Integer) }
      end

      def disallowed
        @disalllowed ||= Set.new
      end
    end
  end
end

File: elixir/lib/app_identity/versions.ex
defmodule AppIdentity.Versions do
  @moduledoc false

  @proof_string_separator ":"

  alias AppIdentity.{App, Validation}

  @version_map %{
    1 => AppIdentity.Versions.V1,
    2 => AppIdentity.Versions.V2,
    3 => AppIdentity.Versions.V3,
    4 => AppIdentity.Versions.V4
  }
  @supported_versions Map.keys(@version_map)

  @doc """
  Returns the algorithm module that matches the provided `version`. Returns
  `{:ok, module}` or `{:error, reason}`.

      iex> AppIdentity.Versions.version(1)
      {:ok, AppIdentity.Versions.V1}

      iex> AppIdentity.Versions.version(5)
      {:error, "unsupported version 5"}
  """
  @spec version(AppIdentity.version()) :: {:ok, module()} | {:error, reason :: String.t()}
  def version(version) do
    case Map.fetch(@version_map, version) do
      {:ok, _} = ok -> ok
      :error -> {:error, "unsupported version #{inspect(version)}"}
    end
  end

  @doc "The list of supported algorithm version numbers."
  @spec supported :: [AppIdentity.version()]
  def supported do
    @supported_versions
  end

  @doc "Generates a nonce appropriate to the provided algorithm version."
  @spec generate_nonce(version :: App.t() | AppIdentity.version()) ::
          {:ok, AppIdentity.nonce()} | {:error, reason :: String.t()}
  def generate_nonce(%{version: app_version}) do
    generate_nonce(app_version)
  end

  def generate_nonce(version) when is_integer(version) do
    case version(version) do
      {:ok, mod} -> mod.generate_nonce()
      error -> error
    end
  end

  @doc "Validates a nonce appropriate to the provided algorithm version."
  @spec validate_nonce(App.t(), AppIdentity.nonce(), AppIdentity.version()) ::
          {:ok, AppIdentity.nonce()} | {:error, reason :: String.t()}
  def validate_nonce(%{config: config}, nonce, version) do
    case version(version) do
      {:ok, mod} -> mod.validate_nonce(nonce, config)
      error -> error
    end
  end

  @doc """
  Creates a digest value for use in a padlock according to the provided algorithm version.
  """
  @spec make_digest(App.t(), AppIdentity.nonce(), AppIdentity.version()) ::
          {:ok, digest :: String.t()} | {:error, reason :: String.t()}
  def make_digest(%{id: id, secret: secret}, nonce, version) do
    case version(version) do
      {:ok, mod} ->
        digest =
          [id, nonce, secret.()]
          |> join()
          |> mod.make_digest()
          |> Base.encode16(case: :upper)

        {:ok, digest}

      error ->
        error
    end
  end

  @doc """
  Tests if the version provided is both supported and has not been explicitly
  disallowed.

  ## Examples

      iex> AppIdentity.Versions.allowed_version(1)
      :ok

      iex> AppIdentity.Versions.allowed_version(1, disallowed: [1])
      {:error, "version 1 has been disallowed"}

      iex> AppIdentity.Versions.allowed_version(5)
      {:error, "unsupported version 5"}

      iex> AppIdentity.Versions.allowed_version(1, disallowed: 5)
      {:error, "error in disallowed version configuration"}
  """
  @spec allowed_version(AppIdentity.version(), [AppIdentity.option()]) ::
          :ok | {:error, reason :: String.t()}
  def allowed_version(version, options \\ [])

  def allowed_version(version, options) when version in @supported_versions do
    case Keyword.get(options, :disallowed) do
      nil ->
        :ok

      [] ->
        :ok

      disallowed when is_list(disallowed) ->
        if version in disallowed do
          {:error, "version #{version} has been disallowed"}
        else
          :ok
        end

      _ ->
        {:error, "error in disallowed version configuration"}
    end
  end

  def allowed_version(version, _options) do
    {:error, "unsupported version #{inspect(version)}"}
  end

  defp join(parts, separator \\ @proof_string_separator) when is_list(parts) do
    parts
    |> Enum.intersperse(separator)
    |> IO.iodata_to_binary()
  end

  defmodule RandomNonce do
    # Generate and validate a randomly generated nonce.
    @moduledoc false

    def generate_nonce do
      {:ok, Base.url_encode64(:crypto.strong_rand_bytes(32))}
    end

    def validate_nonce(nonce, _config) do
      Validation.validate(:nonce, nonce)
    end
  end

  defmodule TimestampNonce do
    # Generate and validate a timestamp nonce.
    @moduledoc false

    def generate_nonce do
      case DateTime.now("Etc/UTC") do
        {:ok, stamp} ->
          {:ok, DateTime.to_iso8601(stamp, :basic)}

        {:error, reason} ->
          {:error, String.replace(Kernel.to_string(reason), "_", " ")}
      end
    end

    def validate_nonce(nonce, config) do
      case Validation.validate(:nonce, nonce) do
        {:ok, nonce} -> parse_and_validate_nonce(nonce, config)
        error -> error
      end
    end

    @default_config %{fuzz: 600}

    defp parse_and_validate_nonce(nonce, config) do
      with {:ok, timestamp} <- parse_timestamp(nonce),
           {:ok, now} <- DateTime.now("Etc/UTC") do
        compare_timestamp(now, timestamp, nonce, config)
      end
    end

    defp compare_timestamp(now, timestamp, nonce, config) do
      config = config || @default_config
      fuzz = Map.get(config, :fuzz) || Map.get(config, "fuzz") || 600
      diff = abs(DateTime.diff(now, timestamp))

      if diff <= fuzz do
        {:ok, nonce}
      else
        {:error, "nonce is invalid"}
      end
    end

    defp parse_timestamp(<<
           year::binary-size(4),
           month::binary-size(2),
           day::binary-size(2),
           "T",
           hour::binary-size(2),
           minute::binary-size(2),
           second::binary-size(2),
           rest::binary
         >>) do
      case DateTime.from_iso8601("#{year}-#{month}-#{day}T#{hour}:#{minute}:#{second}#{rest}") do
        {:ok, timestamp, _offset} -> {:ok, timestamp}
        {:error, reason} -> {:error, String.replace(Kernel.to_string(reason), "_", " ")}
      end
    end

    defp parse_timestamp(_) do
      {:error, "nonce does not look like a timestamp"}
    end
  end

  defmodule V1 do
    @moduledoc false

    defdelegate generate_nonce, to: RandomNonce
    defdelegate validate_nonce(nonce, config), to: RandomNonce

    def make_digest(raw) do
      :crypto.hash(:sha256, raw)
    end
  end

  defmodule V2 do
    @moduledoc false

    defdelegate generate_nonce, to: TimestampNonce
    defdelegate validate_nonce(nonce, config), to: TimestampNonce

    def make_digest(raw) do
      :crypto.hash(:sha256, raw)
    end
  end

  defmodule V3 do
    @moduledoc false

    defdelegate generate_nonce, to: TimestampNonce
    defdelegate validate_nonce(nonce, config), to: TimestampNonce

    def make_digest(raw) do
      :crypto.hash(:sha384, raw)
    end
  end

  defmodule V4 do
    @moduledoc false

    defdelegate generate_nonce, to: TimestampNonce
    defdelegate validate_nonce(nonce, config), to: TimestampNonce

    def make_digest(raw) do
      :crypto.hash(:sha512, raw)
    end
  end
end
